src/main/java/org/folio/fqm/service/EntityTypeFlatteningService.java

```diff
@@ -0,0 +1,242 @@
+      .join(nestedSource.getJoin())
+      .useIdColumns(outerSource == null || Boolean.TRUE.equals(outerSource.getUseIdColumns()));
+    String nestedAlias = newSource.getAlias();
+
+    StringBuilder newAlias = outerSource != null ? new StringBuilder(outerSource.getAlias()) : new StringBuilder();
+    if (keepOriginalAlias) {
+      newAlias.append("_").append(nestedAlias);
+    }
+    log.info("Updating source/columns for db source for original entity type " + originalEntityType.getName());
+    for (EntityTypeColumn oldColumn : originalEntityType.getColumns()) {
+      EntityTypeColumn column = copyColumn(oldColumn);
+      if (column.getSourceAlias().equals(nestedAlias)) {
+        if (outerSource != null) { // temporary, need a better way to do this
+          column.name(outerSource.getAlias() + "." + column.getName());
+          // Only treat column as idColumn if outer source specifies to do so
+          column.isIdColumn(Boolean.TRUE.equals(outerSource.getUseIdColumns()) && Boolean.TRUE.equals(column.getIsIdColumn()));
+          if (!Boolean.TRUE.equals(newSource.getFlattened())) {
+            column.sourceAlias(newAlias.toString());
+          }
+        }
+        updatedColumns.add(column);
+      }
+    }
+
+    if (outerSource != null && !Boolean.TRUE.equals(newSource.getFlattened())) { // TODO: may not need "nestedSource.getJoin() == null"
+      newSource.alias(newAlias.toString());
+      newSource.flattened(true);
+      if (nestedSource.getJoin() == null) {
+        newSource.join(outerSource.getJoin());
+      }
+    }
+    return new Pair<>(newSource, updatedColumns);
+  }
+
+  private List<EntityTypeColumn> finalColumnConversion(EntityType flattenedEntityType) {
+    List<EntityTypeColumn> finalColumns = new ArrayList<>();
+    String toReplace = ":sourceAlias";
+    for (EntityTypeColumn column : flattenedEntityType.getColumns()) {
+      String sourceAlias = "\"" + column.getSourceAlias() + "\"";
+      String valueGetter = column.getValueGetter();
+      String filterValueGetter = column.getFilterValueGetter();
+      valueGetter = valueGetter.replace(toReplace, sourceAlias);
+      if (filterValueGetter != null) {
+        filterValueGetter = filterValueGetter.replace(toReplace, sourceAlias);
+      }
+      column.valueGetter(valueGetter);
+      column.filterValueGetter(filterValueGetter);
+      finalColumns.add(column);
+    }
+    return finalColumns;
+  }
+
+  private EntityTypeColumn copyColumn(EntityTypeColumn column) {
+    try {
+      String json = objectMapper.writeValueAsString(column);
+      return objectMapper.readValue(json, EntityTypeColumn.class);
+    } catch (Exception e) {
+      return new EntityTypeColumn(); // TODO: do something better here
+    }
+  }
+
+  private long countDbSources(EntityType entityType) {
+    return entityType
+      .getSources()
+      .stream()
+      .filter(source -> !Boolean.TRUE.equals(source.getFlattened()) && source.getType().equals("db"))
```

Comment:
Maybe set the type to "flattened" or something and get rid of the `flattened` property? That'd simplify the object model a bit and probably simplify some code outside this method, too, too